File: BinarySearch2D.java
// Time Complexity :O(log n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :none

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        // init low pointer
        int low= 0;
        // init high pointer
        int high= (matrix.length)*(matrix[0].length)-1;
        // we will be treating the 2d array as a one d array to solve the problem
        
        // Iterate while low <=high
        while (low<=high){
            // find mid element
            int mid= low+(high-low)/2;
            // find the row position of mid by dividing the mid by number of columns
            int row=mid/matrix[0].length;
            // find the column position of the mid element by taking mod of mid by number of column
            int col=mid%matrix[0].length;
            
            // if target is mid element return true
            if(target==matrix[row][col]){
                return true;
            }
            // if target is smaller than high, update high
            if (target<=matrix[row][col]){
                high= mid-1;
            }
            // if target is greater than mid, update low
            if (target>matrix[row][col]){
                low= mid+1;
            }
                
        }
        return false;
    }
    
}

File: binary-search-2D.py
## Problem1 
# Search a 2D Matrix(https://leetcode.com/problems/search-a-2d-matrix/)
# // Time Complexity : O(logn)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : yes
# // Any problem you faced while coding this : both the codes has almost same run time. 

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        low =  0 
        m = len(matrix)
        n = len(matrix[0])
        high = m * n - 1
        while low<=high:
            mid = low + int((high - low)/2)
            row = mid // n
            col = mid % n
            if target == matrix[row][col]: 
                return True
            elif target > matrix[row][col]:
                low = mid + 1
            else:
                high = mid - 1
        return False
        # Binary search untill find the row - old code 
        # fr = 0
        # lr = len(matrix) - 1
        # while fr<=lr:
        #     mr = int(fr + (lr - fr)/2)
        #     if matrix[mr][0]<=target and target<=matrix[mr][-1]:
        #         if target in matrix[mr]:
        #             return True
        #         else:
        #             return False
        #     elif target<matrix[mr][0]:
        #         lr = mr - 1
        #     elif target>matrix[mr][-1]:
        #         fr = mr + 1
